File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
  challenger = [0] * (N+2) # 스테이지 별 도전자수 구함
  for stage in stages:
    challenger[stage] += 1

  #스테이지 별 실패한 사용자 수 계싼
  fails = {}
  total = len(stages)

  #각 스테이지를 순회하며, 실패율 계산
  for i in range(1, N+1):
    if challenger[i] == 0: #도전한 사람이 없는 경우, 실패율은 0
      fails[i] = 0
    else:
      fails[i] = challenger[i] / total # 실패율 구함
      total -= challenger[i] # 다음 스테이지 실패율을 구하기 위해 현재 스테이지의 인원을 뺌

  result = sorted(fails, key = lambda x: fails[x], reverse=True)

  return result



# 책 풀이

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
  // 현재 도전중인 스테이지 번호가 각각 있음 
  let challenger = new Array(N + 2).fill(0); // 0 1 2 3 4 5 6 이미 도전중인  

  let fails = new Map(); //실패 확률 작성할 Map
  stages.map((el, idx) => {
    challenger[el] += 1;
  });

  // 실패한거 1단계하면 전체 세고 # 현재 숫자빼면됨
  let ans = []
  let prev = 0;

  let cur_stage = 1;
  while (cur_stage <= N) {
    let percent = challenger[cur_stage] / (stages.length - prev);
    if (challenger[cur_stage] == 0) {
      ans.push([cur_stage, 0])
    } else {
      ans.push([cur_stage, percent])
    }
    prev += challenger[cur_stage];
    cur_stage++;
  }


  ans.sort((a, b) => {
    if (a[1] == b[1]) {
      return a[0] - b[0] //음수면 b가 앞에  
    } else {
      return b[1] - a[1]
    }
  })

  let result = ans.map((el) => el[0])

  return result;
}


// 0 1 3 2 1 1